gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordHistoryService.java

```diff
@@ -0,0 +1,135 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.service.impl;
+
+import io.gravitee.am.common.audit.EventType;
+import io.gravitee.am.identityprovider.api.User;
+import io.gravitee.am.model.PasswordHistory;
+import io.gravitee.am.model.PasswordSettings;
+import io.gravitee.am.model.ReferenceType;
+import io.gravitee.am.repository.management.api.PasswordHistoryRepository;
+import io.gravitee.am.service.AuditService;
+import io.gravitee.am.service.authentication.crypto.password.PasswordEncoder;
+import io.gravitee.am.service.exception.PasswordHistoryException;
+import io.gravitee.am.service.reporter.builder.AuditBuilder;
+import io.gravitee.am.service.reporter.builder.management.UserAuditBuilder;
+import io.reactivex.Single;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.List;
+
+import static java.util.Comparator.comparing;
+
+/**
+ * Service providing password history.
+ */
+@Component
+public class PasswordHistoryService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PasswordHistoryService.class);
+    private final PasswordHistoryRepository repository;
+    private final AuditService auditService;
+    private final PasswordEncoder passwordEncoder;
+
+    @Autowired
+    public PasswordHistoryService(@Lazy PasswordHistoryRepository repository, AuditService auditService, PasswordEncoder passwordEncoder) {
+        this.repository = repository;
+        this.auditService = auditService;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    /**
+     * Checks that a password has not already been used in the allotted history and adds it if it has not.
+     * Returns a {@link PasswordHistory} instance on completion.
+     *
+     * @param referenceType    the reference type (DOMAIN, APPLICATION)
+     * @param referenceId      id of the reference
+     * @param userId           id of user for this password
+     * @param password         the password to add
+     * @param principal        user performing this action
+     * @param passwordSettings domain/application password settings
+     * @return Single containing {@link PasswordHistory} or an error if the password was already in the history.
+     */
+    public Single<PasswordHistory> addPasswordToHistory(ReferenceType referenceType, String referenceId, String userId, String password, User principal, PasswordSettings passwordSettings) {
+        LOGGER.debug("Adding password history entry for user {}", userId);
+        if (passwordSettings == null || !passwordSettings.isPasswordHistoryEnabled()) {
+            return Single.error(new IllegalArgumentException("Password history disabled."));
```

Comment:
Should not return an error in this case in my opinion, but return an empty value.
As `Single<>` can't accept null value, the return type for this method should be `Maybe`.

Otherwise you have to manage the Error in the upstream flow during the ResetPassword to avoid blocking the resetPassword action if the history is disabled. (see postman test error)

Reply:
Yes I thought the same when I saw the failing jest tests.